wrapper/pre_copy.py

```diff
@@ -391,46 +408,66 @@ def copy(self, vm, final, keepalive=None):
         copy.status = 'Connected'
         STATE.write()
 
-        self.get_extents(vm, final)
-        if len(self.extents) == 0:
-            copy.to_copy = 0
-            copy.copied = 0
-            copy.end_time = time.time()
-            if not final and len(self.copies) > 2:
-                self.copies[-2].path = None
-                self.copies[-2].change_id = None
-                self.copies[-3].snapshot = None
-            return
-
-        nbd_handle = self.nbd.NBD()
-        nbd_handle.add_meta_context('base:allocation')
-        nbd_handle.connect_uri(nbd_uri_from_unix_socket(self.sock))
-        fd = os.open(self.local_path, os.O_WRONLY)
+        fd = None
+        nbd_handle = None
 
         try:
+            self.get_extents(vm, final)
+            if len(self.extents) == 0:
+                copy.to_copy = 0
+                copy.copied = 0
+                copy.end_time = time.time()
+                copy.status = 'Copied'
+                if final:
+                    if len(self.copies) > 1:
+                        self.copies[-2].snapshot = None
+                elif len(self.copies) > 2:
+                    self.copies[-2].path = None
+                    self.copies[-2].change_id = None
+                    self.copies[-3].snapshot = None
+                return
+
+            nbd_handle = self.nbd.NBD()
+            nbd_handle.add_meta_context('base:allocation')
+            nbd_handle.connect_uri(nbd_uri_from_unix_socket(self.sock))
+            fd = os.open(self.local_path, os.O_WRONLY)
+
             copy.status = 'Copying'
             STATE.write()
             self._copy_all(nbd_handle, fd, final, keepalive)
             copy.status = 'Copied'
             copy.end_time = time.time()
-            if not final and len(self.copies) > 2:
+            if final:
+                if len(self.copies) > 1:
+                    self.copies[-2].snapshot = None
+            elif len(self.copies) > 2:
                 self.copies[-2].path = None
                 self.copies[-2].change_id = None
                 self.copies[-3].snapshot = None
             STATE.write()
+        except (STATE.pre_copy.pyvmomi.vim.fault.VimFault,
+                STATE.pre_copy.nbd.Error) as exc:
+            if final:
+                # TODO: Fail?  Retry?  What function should be handling that?
+                error('Error during final copy iteration', exception=True)
+                raise
```

Comment:
My understanding was that when the final copy starts and because the VM is stopped and we are in the middle of downtime, we need this to either complete or fail, as fast as possible.

If there is *anything* that starts working with *any* resource from the VM, then that's a bug somewhere else. The wrapper should be expected to have full control of the VM if `.status != "Paused"`. Otherwise we'll be chasing thousands of corner cases because "working in every possible scenario" is impossible to maintain.

Reply:
Ok. Fine with this approach. We can still revisit later if we have users complaints ;)